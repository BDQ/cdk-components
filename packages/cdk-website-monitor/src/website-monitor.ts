import * as path from 'path';
import { Construct, CfnOutput } from '@aws-cdk/core';
import {
    Dashboard,
    CfnDashboard,
    IWidget,
    GraphWidget,
    TextWidget,
} from '@aws-cdk/aws-cloudwatch';
import { Function, Runtime, Code } from '@aws-cdk/aws-lambda';
import { Rule, Schedule } from '@aws-cdk/aws-events';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import { PolicyStatement } from '@aws-cdk/aws-iam';

import { SuccessMetric } from './success-metric';
import { LatencyMetric } from './latency-metric';

function linkForDashboard(dashboard: Dashboard): string {
    const cfnDashboard = dashboard.node.defaultChild as CfnDashboard;
    return `https://console.aws.amazon.com/cloudwatch/home?region=${dashboard.stack.region}#dashboards:name=${cfnDashboard.ref}`;
}

interface SectionOptions {
    readonly links?: {
        readonly title: string;
        readonly url: string;
    }[];
}

export interface WatchServiceProps {
    readonly serviceName: string;
    readonly url: string;
    readonly expression: string;
}

export class WebsiteMonitor extends Construct {
    private readonly dashboard: Dashboard;

    private readonly function: Function;

    public constructor(scope: Construct, id: string) {
        super(scope, id);

        this.dashboard = new Dashboard(this, 'Dashboard');

        this.function = new Function(this, 'Function', {
            runtime: Runtime.NODEJS_10_X,
            code: Code.asset(
                path.join(__dirname, '..', 'lambda', 'bundle.zip'),
            ),
            handler: 'lib/index.handler',
        });

        this.function.addToRolePolicy(
            new PolicyStatement({
                actions: ['cloudwatch:PutMetricData'],
                resources: ['*'],
            }),
        );

        new CfnOutput(this, 'WebsiteMonitorDashboard', {
            value: linkForDashboard(this.dashboard),
        });
    }

    private addSection(title: string, options: SectionOptions = {}): void {
        const markdown = [
            `# ${title}`,
            (options.links || [])
                .map((link): string => `[button:${link.title}](${link.url})`)
                .join(' | '),
        ];

        this.addWidgets(
            new TextWidget({
                width: 24,
                markdown: markdown.join('\n'),
            }),
        );
    }

    private addWidgets(...widgets: IWidget[]): void {
        this.dashboard.addWidgets(...widgets);
    }

    public watchService(props: WatchServiceProps): void {
        const { serviceName, url, expression } = props;

        this.addSection(serviceName, { links: [{ title: serviceName, url }] });

        const successMetric = new SuccessMetric(serviceName);
        const latencyMetric = new LatencyMetric(serviceName);

        this.addWidgets(
            new GraphWidget({
                title: `Success`,
                width: 6,
                left: [successMetric],
            }),
            new GraphWidget({
                title: `Latency`,
                width: 6,
                left: [latencyMetric],
            }),
        );

        const scheduler = new Rule(this, 'Scheduler', {
            schedule: Schedule.expression(expression),
        });

        scheduler.addTarget(new LambdaFunction(this.function));
    }
}
