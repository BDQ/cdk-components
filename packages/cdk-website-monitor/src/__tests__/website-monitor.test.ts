import { Stack } from '@aws-cdk/core';
import '@cloudcomponents/jest-cdk-snapshot';
import { WebsiteMonitor } from '../website-monitor';

describe('cdk-website-monitor: website-monitor', (): void => {
    it('snapshot', (): void => {
        const stack = new Stack();

        const websiteMonitor = new WebsiteMonitor(stack, 'WebsiteMonitor');

        websiteMonitor.watchService({
            serviceName: 'cloudcomponents',
            url: 'https://www.cloudcomponents.org',
            expression: 'rate(2 minutes)',
        });

        expect(stack).toMatchCdkSnapshot();
    });
});
