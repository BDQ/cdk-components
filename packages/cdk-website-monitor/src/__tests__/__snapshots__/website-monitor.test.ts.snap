// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-website-monitor: website-monitor snapshot 1`] = `
Object {
  "Outputs": Object {
    "WebsiteMonitorWebsiteMonitorDashboard845E6E19": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://console.aws.amazon.com/cloudwatch/home?region=",
            Object {
              "Ref": "AWS::Region",
            },
            "#dashboards:name=",
            Object {
              "Ref": "WebsiteMonitorDashboard74214FA0",
            },
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "WebsiteMonitorFunctionCodeArtifactHashF83E0178": Object {
      "Description": "Artifact hash for asset \\"WebsiteMonitor/Function/Code\\"",
      "Type": "String",
    },
    "WebsiteMonitorFunctionCodeS3BucketF763597F": Object {
      "Description": "S3 bucket for asset \\"WebsiteMonitor/Function/Code\\"",
      "Type": "String",
    },
    "WebsiteMonitorFunctionCodeS3VersionKey8A7FB0F8": Object {
      "Description": "S3 key for asset version \\"WebsiteMonitor/Function/Code\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "WebsiteMonitorDashboard74214FA0": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":2,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"# cloudcomponents\\\\n[button:cloudcomponents](https://www.cloudcomponents.org)\\"}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":2,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Success\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"CloudComponents/WebsiteMonitor\\",\\"Success\\",\\"ServiceName\\",\\"cloudcomponents\\",{\\"yAxis\\":\\"left\\",\\"period\\":300,\\"stat\\":\\"Average\\"}]],\\"annotations\\":{\\"horizontal\\":[]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":2,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Latency\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"CloudComponents/WebsiteMonitor\\",\\"Latency\\",\\"ServiceName\\",\\"cloudcomponents\\",{\\"yAxis\\":\\"left\\",\\"period\\":300,\\"stat\\":\\"Average\\"}]],\\"annotations\\":{\\"horizontal\\":[]},\\"yAxis\\":{}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "WebsiteMonitorFunction910AC36F": Object {
      "DependsOn": Array [
        "WebsiteMonitorFunctionServiceRoleDefaultPolicyC501829D",
        "WebsiteMonitorFunctionServiceRole174679C0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "WebsiteMonitorFunctionCodeS3BucketF763597F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "WebsiteMonitorFunctionCodeS3VersionKey8A7FB0F8",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "WebsiteMonitorFunctionCodeS3VersionKey8A7FB0F8",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lib/index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebsiteMonitorFunctionServiceRole174679C0",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebsiteMonitorFunctionAllowEventRuleWebsiteMonitorSchedulerF7C246B05CE56A31": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WebsiteMonitorFunction910AC36F",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "WebsiteMonitorScheduler4DC65597",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebsiteMonitorFunctionServiceRole174679C0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "lambda.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebsiteMonitorFunctionServiceRoleDefaultPolicyC501829D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebsiteMonitorFunctionServiceRoleDefaultPolicyC501829D",
        "Roles": Array [
          Object {
            "Ref": "WebsiteMonitorFunctionServiceRole174679C0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebsiteMonitorScheduler4DC65597": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(2 minutes)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "WebsiteMonitorFunction910AC36F",
                "Arn",
              ],
            },
            "Id": "WebsiteMonitorFunction32A41AC6",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
